Refine(): Added new function for MatrixElement of symmetric matrix

arooshiverma opened the pull request:
Added feature for giving corresponding equal above-diagonal elements for input of elements below diagonal.

Example:

>>>x=MatrixSymbol('x',5,5)
>>>refine(x[0,1],Q.symmetric(x))
x[0,1]
>>>refine(x[1,0],Q.symmetric(x))
x[0,1]

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Added a new function refine_matrixelement(expression,assumptions) to refine.py

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
  * added new function in refine
<!-- END RELEASE NOTES -->

arooshiverma committed:
Refine(): Added new function for MatrixElement of symmetric matrix

Added feature for giving corresponding equal above-diagonal elements for input of elements below diagonal.

Example:

>>>x=MatrixSymbol('x',5,5)
>>>refine(x[0,1],Q.symmetric(x))
x[0,1]
>>>refine(x[1,0],Q.symmetric(x))
x[0,1]

Abdullahjavednesar commented:
Please add tests.

ShubhamKJha commented on sympy/assumptions/refine.py:
```diff
@@ -239,6 +240,16 @@ def refine_Relational(expr, assumptions):
+def refine_matrixelement(expr,assumptions):
+
+
+
```
Remove these extra spaces.

ShubhamKJha commented on sympy/assumptions/refine.py:
```diff
@@ -239,6 +240,16 @@ def refine_Relational(expr, assumptions):
+
+    x=expr.parent
+    if (ask(Q.symmetric(x),assumptions)==True):
+        if(expr.args[1]+expr.args[2]>2*expr.args[2]):
```
As @asmeurer  said `(i - j).could_extract_minus_sign()` could better handle symbolic expressions.

arooshiverma replied:
Could you elaborate this a little?

ShubhamKJha replied:
The way I see, it returns True if `i>j` when `i` and `j` are both numbers (in SymPy). But in case of symbols it checks for the `canonical` form of the indices ( both being symbols) hence it's better to use this.
Also isn't the expression equivalent to `expr.args[1] > expr.args[2]` ?

arooshiverma replied:
> Also isn't the expression equivalent to `expr.args[1] > expr.args[2]` ?

Yeah just noticed this. Will correct it.

ShubhamKJha replied:
> `(i - j).could_extract_minus_sign()` I believe you should use it.

ShubhamKJha reviewed (commented):
Always add the issue number you are solving in the PR and follow PEP 8.

arooshiverma commented:
> Please add tests.

Yeah I will

arooshiverma committed:
Refine() for MatrixElements

-Extra spaces in refine_matrixelement() definition removed
-Added new test in test_refine: test_matrixelement()

sylee957 commented:
Reference #16234

sylee957 commented on sympy/assumptions/refine.py:
```diff
@@ -1,5 +1,6 @@
 from __future__ import print_function, division
 
+#from sympy.matrices import MatrixExpr
```
It may be deleted if not needed

arooshiverma replied:
But if we delete it, it doesn't pass a test.

ShubhamKJha replied:
Travis tests were not passing because of some network error. Presence of an unimportant comment can do nothing to the tests.

arooshiverma replied:
Sorry I got confused. I didn't realise that was a comment. Got it confused with something else.

sylee957 reviewed (commented):
Looks good.
But we may have to follow the coding convention to add an extra spacing after comma like
`def refine_matrixelement(expr, assumptions):`
or 
`x[0, 1]`

arooshiverma committed:
Refine(): Modified according to coding convention

Added space after comma

arooshiverma committed:
Refine(): corrected the condition in refine_matrixelement()

arooshiverma committed:
refine(): Deleted unwanted comment from refine.py

sylee957 committed:
Coding convention cleanups

sylee957 committed:
Cleanup some brackets

sylee957 committed:
Rename matrixelement to MatrixElement

arooshiverma commented:
Can this be merged?

czgdp1807 commented:
@arooshiverma It looks like a good work. Would you like to complete this PR? If so, then please resolve the conflicts and address the above comments. Thanks for the contributions.

faze-geek commented:
Could be closed via #18681 .
